adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/LoraProtocolAdapter.java

```diff
@@ -0,0 +1,636 @@
+import org.eclipse.hono.adapter.http.HttpProtocolAdapterProperties;
+import org.eclipse.hono.adapter.http.X509AuthHandler;
+import org.eclipse.hono.adapter.lora.providers.LoraProvider;
+import org.eclipse.hono.adapter.lora.providers.LoraProviderMalformedPayloadException;
+import org.eclipse.hono.adapter.lora.providers.LoraUtils;
+import org.eclipse.hono.auth.Device;
+import org.eclipse.hono.client.Command;
+import org.eclipse.hono.client.CommandContext;
+import org.eclipse.hono.client.CommandResponse;
+import org.eclipse.hono.client.MessageConsumer;
+import org.eclipse.hono.service.auth.device.HonoClientBasedAuthProvider;
+import org.eclipse.hono.service.auth.device.TenantServiceBasedX509Authentication;
+import org.eclipse.hono.service.auth.device.UsernamePasswordAuthProvider;
+import org.eclipse.hono.service.auth.device.X509AuthProvider;
+import org.eclipse.hono.service.http.HttpUtils;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.CredentialsObject;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import io.opentracing.References;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.contrib.vertx.ext.web.TracingHandler;
+import io.opentracing.noop.NoopSpan;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Future;
+import io.vertx.core.MultiMap;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.HttpVersion;
+import io.vertx.core.json.DecodeException;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.client.HttpResponse;
+import io.vertx.ext.web.handler.ChainAuthHandler;
+
+/**
+ * A Vert.x based Hono protocol adapter for receiving HTTP push messages from and sending commands to LoRa backends.
+ */
+public final class LoraProtocolAdapter extends AbstractVertxBasedHttpProtocolAdapter<HttpProtocolAdapterProperties> {
+
+    private static final String ERROR_MSG_MISSING_OR_UNSUPPORTED_CONTENT_TYPE = "missing or unsupported content-type";
+    private static final Logger LOG = LoggerFactory.getLogger(LoraProtocolAdapter.class);
+    private static final String LORA_COMMAND_CONSUMER_DEVICE_ID = "lora";
+    private static final int LORA_COMMAND_CONSUMER_RETRY_INTERVAL = 2_000;
+    private static final String TAG_LORA_DEVICE_ID = "lora_device_id";
+    private static final String TAG_LORA_PROVIDER = "lora_provider";
+    private static final String JSON_MISSING_REQUIRED_FIELDS = "JSON Body does not contain required fields";
+    private static final String INVALID_PAYLOAD = "Invalid payload";
+    @Autowired
+    private final List<LoraProvider> loraProviders = new ArrayList<>();
```

Comment:
please add a setter for this property and move the annotation to the setter to reduce coupling with Spring field injection.